src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+                for (datum, other_spec) in b {
+                    let spec = a.entry(datum).or_insert_with(ResultSpec::anything);
+                    *spec = spec.clone().intersect(other_spec);
+                }
+                Values::Nested(a)
+            }
+            (Values::All, v) => v,
+            (v, Values::All) => v,
+            (Values::Nested(_), Values::Within(_, _)) => Values::Empty,
+            (Values::Within(_, _), Values::Nested(_)) => Values::Empty,
+        }
+    }
+
+    fn may_contain(&self, value: Datum<'a>) -> bool {
+        match self {
+            Values::Empty => false,
+            Values::Within(min, max) => *min <= value && value <= *max,
+            Values::All => true,
+            Values::Nested(field_map) => match value {
+                Datum::Map(datum_map) => {
+                    datum_map
+                        .iter()
+                        .all(|(key, val)| match field_map.get(&key.into()) {
+                            None => true,
+                            Some(nested) => nested.may_contain(val),
+                        })
+                }
+                _ => false,
+            },
+        }
+    }
+}
+
+/// An approximation of the set of values an expression might have, including whether or not it
+/// might be null or an error value. This is generally an _overapproximation_, in the sense that
+/// [ResultSpec::may_contain] may return true even if the argument is not included in the set.
+/// (However, it should never return false when the value _is_ included!)
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ResultSpec<'a> {
+    /// True if the expression may evaluate to [Datum::Null].
+    nullable: bool,
+    /// True if the expression may evaluate to an error.
+    fallible: bool,
+    /// The range of possible (non-null) values that the expression may evaluate to.
+    values: Values<'a>,
+}
+
+impl<'a> ResultSpec<'a> {
+    /// No results match this spec. (For example, an empty table.)
+    pub fn nothing() -> Self {
+        ResultSpec {
+            nullable: false,
+            fallible: false,
+            values: Values::Empty,
+        }
+    }
+
+    /// Every result matches this spec.
+    pub fn anything() -> Self {
+        ResultSpec {
+            nullable: true,
+            fallible: true,
+            values: Values::All,
+        }
+    }
+
+    /// A spec that only matches null.
+    pub fn null() -> Self {
+        ResultSpec {
+            nullable: true,
+            ..Self::nothing()
+        }
+    }
+
+    /// A spec that only matches error values.
+    pub fn fails() -> Self {
+        ResultSpec {
+            fallible: true,
+            ..Self::nothing()
+        }
+    }
+
+    /// A spec that matches all values of a given type.
+    pub fn has_type(col: &ColumnType, fallible: bool) -> ResultSpec<'a> {
+        let values = match &col.scalar_type {
+            ScalarType::Bool => Values::Within(Datum::False, Datum::True),
```

Comment:
Any reason not to do similarly for numeric types that have mins and maxes?

Reply:
Only that it hasn't been necessary yet, though I agree that it would be valid and improve the bounds.

(It's especially useful for booleans because of the boolean-specific case in `flat_map`, incidentally.)